journals/week5.md

```diff
@@ -0,0 +1,186 @@
+# Week 5: Prototyping and Jobtap 
+
+### Refactoring Fripp Code
+
+Since last week, I made some small changes to the implementation in `fripp.c`. First, I added the socket and the packet buffer to the `fripp_ctx`.
+
+```c
+struct fripp_ctx {
+	flux_t *h;
+	struct sockaddr_in si_server;
+	int sock;
+	char *buf;
+	int buf_len;
+};
+```
+
+In turn, that means that the socket is now opened only once per context (when it's being created in `fripp_ctx_create`) and closed once upon destruction of the context, and there is no longer a fixed-size internal buffer for each send (`fripp_format_send`). Instead the buffer in the context is dynamically sized and resized if needed with the idea that most packets being sent will be the same length for a while. One last notable change is that within `fripp_format_send` there is now a second `va_list` (a copy) that's used when the buffer needs to be resized. This was because sometimes, and I'm not sure why it didn't happen every time, upon resizing the buffer the second `vsnprintf` would segfault.
```

Comment:
Those va_list bugs can be tricky.  Good sleuthing!

From stdarg(3):
> Each call to `va_arg()` modifies ap  so  that the  next  call returns the next argument. 

and

> If  there  is  no  next argument, or if type is not compatible with the type of the actual next argument (as promoted according to the  default argument promotions), random errors will occur.

So the first call to `vsnprintf()` walks through all the args with `va_arg()`, so in the second call to `vsnprintf()`, there is no next argument and "random errors occur".

C really takes care of you as a programmer...said no one, ever!